pkg/lib/serverTime.js

```diff
@@ -439,63 +610,142 @@
+                                    icon={ <CloseIcon /> } className="remove-server"
+                                    isDisabled={servers.length == 1 && servers[0] == ""} />
                         </td>
                     </tr>))
                 }
             </tbody>
         </table>);
 
+    return (
+        <Validated errors={errors} error_key="ntp_servers">
+            <div id="systime-ntp-servers-row">
+                <div id="systime-ntp-servers">
+                    { server_rows }
+                </div>
+            </div>
+        </Validated>);
+};
+
+const StaticServerList = ({ servers, change, map }) => {
+    function label(name) {
+        if (map && map[name])
+            return map[name];
+        else
+            return name;
+    }
+    return <>
+        { servers.map(s => <Checkbox key={s.name}
+                                            id={"cb-" + s.name}
+                                            label={label(s.name)}
+                                            isChecked={s.enabled}
+                                            onChange={val => { s.enabled = val; change() }} />)
+        }
+    </>;
+};
+
+const TimesyncdConfig = ({ state, errors, change }) => {
+    return <>
+        <Checkbox id="cb-specific"
+                         label={_("Use specific time servers")}
+                         isChecked={state.enabled}
+                         onChange={val => { state.enabled = val; change() }} />
+        { state.enabled &&
+        <DynamicServerList servers={state.servers} errors={errors} change={change} />
+        }
+    </>;
+};
+
+const chrony_sources_map = {
+    "/run/chrony-dhcp": _("NTP servers received via DHCP"),
+};
+
+const ChronydConfig = ({ state, errors, change }) => {
+    return <>
+        { state.standard.length > 0 &&
+        <>
+            <div id="standard-servers">{_("Standard servers:")}</div>
+            <StaticServerList servers={state.standard} change={change} />
+        </>
+        }
+        { state.sources.length > 0 &&
+        <>
+            <div id="standard-sources">{_("Additional sources of servers:")}</div>
+            <StaticServerList servers={state.sources} change={change} map={chrony_sources_map} />
+        </>
+        }
+        <div id="additional-sources">{_("Additional servers:")}</div>
+        <DynamicServerList servers={state.extra} errors={errors} change={change} />
+    </>;
+};
+
+const NTPConfig = ({ state, errors, change }) => {
+    if (state.backend == "timesyncd")
+        return <TimesyncdConfig state={state} errors={errors} change={change} />;
+    if (state.backend == "chronyd")
+        return <ChronydConfig state={state} errors={errors} change={change} />;
+};
+
+function validate_ntp_config(state, errors) {
+    if (state.enabled && state.servers.filter(s => !!s).length == 0)
+        errors.ntp_servers = _("Need at least one NTP server");
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17992-20230125-145211-18fde629-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)